Source/WebCore/html/PermissionsPolicy.h

```diff
@@ -63,45 +64,36 @@ class PermissionsPolicy {
         XRSpatialTracking,
 #endif
         PrivateToken,
+        Invalid
     };
-    bool allows(Type, const SecurityOriginData&) const;
-
-    struct AllowRule {
-        enum class Type { All, None, List };
-        Type type { Type::List };
-        HashSet<SecurityOriginData> allowedList;
-    };
+    bool allows(Feature, const SecurityOriginData&) const;
 
 private:
-    static PermissionsPolicy parse(Document&, const HTMLIFrameElement*, StringView);
+    PermissionsPolicy(Document&, const HTMLIFrameElement*);
 
-    AllowRule m_cameraRule;
-    AllowRule m_microphoneRule;
-    AllowRule m_speakerSelectionRule;
-    AllowRule m_displayCaptureRule;
-    AllowRule m_gamepadRule;
-    AllowRule m_geolocationRule;
-    AllowRule m_paymentRule;
-    AllowRule m_syncXHRRule;
-    AllowRule m_fullscreenRule;
-    AllowRule m_webShareRule;
-    AllowRule m_screenWakeLockRule;
+    // https://w3c.github.io/webappsec-permissions-policy/#allowlists
+    class Allowlist {
+    public:
+        Allowlist() = default;
+        enum AllowAllOriginsTag { AllowAllOrigins };
+        explicit Allowlist(AllowAllOriginsTag) : m_origins(AllowAllOrigins) { }
+        explicit Allowlist(const SecurityOriginData& origin): m_origins(HashSet<SecurityOriginData> { origin }) { }
+        void addOrigin(const SecurityOriginData&);
+        bool matches(const SecurityOriginData&) const;
 
-#if ENABLE(DEVICE_ORIENTATION)
-    AllowRule m_gyroscopeRule;
-    AllowRule m_accelerometerRule;
-    AllowRule m_magnetometerRule;
-#endif
-#if ENABLE(WEB_AUTHN)
-    AllowRule m_publickeyCredentialsGetRule;
-#endif
-#if ENABLE(WEBXR)
-    AllowRule m_xrSpatialTrackingRule;
-#endif
-    AllowRule m_privateTokenRule;
+    private:
+        std::variant<HashSet<SecurityOriginData>, AllowAllOriginsTag> m_origins;
+    };
+    Ref<SecurityOrigin> declaredOrigin(const HTMLIFrameElement&) const;
+    Allowlist parseAllowlist(StringView, const SecurityOriginData& containerOrigin, const SecurityOriginData& targetOrigin, bool useStarAsDefaultAllowlistValue);
+
+    using PolicyDirective = HashMap<Feature, Allowlist, IntHash<Feature>, WTF::StrongEnumHashTraits<Feature>>;
+    PolicyDirective parsePolicyDirective(StringView, const SecurityOriginData& containerOrigin, const SecurityOriginData& targetOrigin, bool useStarAsDefaultAllowlistValue);
+
+    PolicyDirective m_effectivePolicy;
```

Comment:
This is more a PermissionsPolicy, a PolicyDirective would be a single entry, not a HashMap.
I would tend to rename PolicyDirective to PolicyDirectives and m_effectivePolicy to m_policies.

Reply:
The spec says "A policy directive is an [ordered map](https://infra.spec.whatwg.org/#ordered-map), mapping [policy-controlled features](https://w3c.github.io/webappsec-permissions-policy/#policy-controlled-feature) to corresponding [allowlists](https://w3c.github.io/webappsec-permissions-policy/#allowlist) of origins."

Reply:
Ah I see, then it is fine.
I was looking at https://w3c.github.io/webappsec-permissions-policy/#serialized-policy-directive where it is a single entry.
I guess the spec is not entirely consistent. It is fine then.